### ISSUE TITLE
Better dotnet tool integration

### ISSUE THREAD

**Issue by gitfool** [cake-build/cake#2685]
It would be great to have better dotnet tool integration, especially now that we have [local tools in .NET Core 3.0](https://andrewlock.net/new-in-net-core-3-local-tools/#local-tools-in-net-core-3-0). I've been pondering what a _minimal_ Cake build could look like. My [BoardGameGeek.Dungeon](https://github.com/gitfool/BoardGameGeek.Dungeon) repo can serve as the use case.

Current Cake related files:
- [cake.ps1](https://github.com/gitfool/BoardGameGeek.Dungeon/blob/master/cake.ps1) bootstraps Cake on Windows
- [cake.sh](https://github.com/gitfool/BoardGameGeek.Dungeon/blob/master/cake.sh) bootstraps Cake on Linux
- [tools/packages.config](https://github.com/gitfool/BoardGameGeek.Dungeon/blob/master/tools/packages.config) locks Cake version

For [Azure Pipelines](https://github.com/gitfool/BoardGameGeek.Dungeon/blob/master/.azure-pipelines.yml#L30) and [GitHub Actions](https://github.com/gitfool/BoardGameGeek.Dungeon/blob/master/.github/workflows/build.yml#L28) builds I avoid the Cake bootstrap by using a [Docker image](https://github.com/gitfool/cake-docker/blob/master/Dockerfile#L10) with the Cake tool already installed. However, all builds still have to bootstrap Cake _modules_ in order to use dotnet tools in the Cake build itself, one of the most common being [GitVersion](https://github.com/GitTools/GitVersion).

BoardGameGeek.Dungeon is built using [Cake.Dungeon](https://github.com/gitfool/Cake.Dungeon), and in the depths lives the [bootstrap.cake](https://github.com/gitfool/Cake.Dungeon/blob/master/scripts/bootstrap.cake) script with the following:
```c#
#module nuget:?package=Cake.DotNetTool.Module&version=0.4.0
...
#tool dotnet:?package=GitVersion.Tool&version=5.1.2
```
This is the first barrier to diet Cake nirvana as it requires Cake to be invoked twice, first to install the `Cake.DotNetTool.Module` to extend Cake to support `#tool`, then again for the actual build, using something like the following:
```
cake --bootstrap && cake
```
Not too bad, although more verbose in [PowerShell](https://github.com/PowerShell/PowerShell) until v7 drops next year with long overdue support for [pipeline chain operators](https://github.com/PowerShell/PowerShell/releases/tag/v7.0.0-preview.5), but really it's surely about time this module moved into Cake core, and when it does we'll want it to play nice with dotnet local tools. (The two step bootstrap will still be required if you're using other Cake modules, but I'd argue that's much less common, and using dotnet [local] tools is standard now, especially if writing cross-platform builds.)

Anyway, given support for dotnet local tools, I would be able to delete the above Cake related files and replace them with a single `dotnet-tools.json` manifest file:
```json
{
  "version": 1,
  "isRoot": true,
  "tools": {
    "cake.tool": {
      "version": "0.36.0",
      "commands": [
        "dotnet-cake"
      ]
    }
  }
}
```
Then to invoke the build on a clean machine I'd use the following:
```
dotnet tool restore && dotnet cake --bootstrap && dotnet cake
```
... or when the dotnet tool module is moved to Cake core:
```
dotnet tool restore && dotnet cake
```
... and when the Cake dotnet tool is already installed (on a local dev machine):
```
dotnet cake
```
Also, for this to integrate better with dotnet _local_ tools, the `#tool` directive should use the local tool directly, presumably invoking `dotnet <tool>`, rather than extracting and copying what was previously a global tool into the Cake tools directory alongside addins and modules. (I'm hoping such a change would also avoid the quirk resulting from building the same directory in Windows and then WSL or Docker with a host mount, or vice versa, where you first need to explicitly delete any such copied dotnet tools which would be for the wrong platform.)

So in summary, a minimal Cake build would simply leverage dotnet local tools to bootstrap Cake itself, and given the dotnet tool module was moved to Cake core, a common case requiring an extra module bootstrap step could most likely be avoided, and the `#tool` directive would use the dotnet local tool directly, rather than copying a platform specific executable into its own tools directory.

**Comment by devlead** [comment]
Good feedback. 

In 1.0 bootstrapping won't need an extra run, believe there's already a PR for that.

Integrating the `Cake.DotNetTool.Module` into Cake is on the table for discussion, IMO it makes sense.

Nothing should stop you from using local tools and tools manifest already today to boostrap tools, Cake tool resolution should find them. But using `#tool` lets you define all build dependencies except cake in one place.

**Comment by pascalberger** [comment]
Implicit bootstrapping of modules has been implemented with https://github.com/cake-build/cake/issues/2833 for Cake 1.0.0.

Documentation on website has also been updated to make .NET Tool the recommended runner: https://cakebuild.net/docs/getting-started/running-cake-scripts

@gitfool Is there anything else or can this issue be closed?

**Comment by gitfool** [comment]
@pascalberger I think this issue can be closed and a couple of new issues created, if agreed and not already existing, focusing on:
* Integrating `Cake.DotNetTool.Module` into Cake core
> Integrating the Cake.DotNetTool.Module into Cake is on the table for discussion, IMO it makes sense.

This is the [most downloaded](https://www.nuget.org/packages?packagetype=&sortby=relevance&q=cake+module&prerel=False) Cake module, which is not surprising given the appeal and utility of portable .NET [Core] tools, and should be built-in to Cake core, unless I'm missing some advantages to having it in a separate module.
* Integrating with dotnet `local` tools
> Also, for this to integrate better with dotnet local tools, the #tool directive should use the local tool directly, presumably invoking dotnet <tool>, rather than extracting and copying what was previously a global tool into the Cake tools directory alongside addins and modules. (I'm hoping such a change would also avoid the quirk resulting from building the same directory in Windows and then WSL or Docker with a host mount, or vice versa, where you first need to explicitly delete any such copied dotnet tools which would be for the wrong platform.)
